sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+            msg_t msg = { .type = DTLS_EVENT_CONNECTED };
+            msg_send(&msg, _auth_waiting_thread);
+        } else if (prev_state == NO_SPACE) {
+            /* No space in session management. Should not happen. If it occurs,
+            we lost track of sessions stored in tinydtls internally */
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
+        }
+
+        /* If not enough session slots left: set timeout to free up session */
+        uint8_t minimum_free = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS;
+        if (dsm_get_num_available_slots() < minimum_free)
+        {
+            uint32_t timeout = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS_TIMEOUT_USEC;
+            event_callback_init(&_dtls_session_free_up_tmout_cb,
+                                _dtls_free_up_session, NULL);
+            event_timeout_init(&_dtls_session_free_up_tmout, &_queue,
+                               &_dtls_session_free_up_tmout_cb.super);
+            event_timeout_set(&_dtls_session_free_up_tmout, timeout);
+        }
+    }
+
+    if (type & SOCK_ASYNC_CONN_FIN) {
+        if (sock_dtls_get_event_session(sock, &socket.ctx_dtls_session)) {
+            dsm_remove(sock, &socket.ctx_dtls_session);
+        } else {
+            puts("gcoap: A session was closed, but the corresponding session " \
+            "could not be retrieved from the socket!");
+            return;
+        }
+        sock_udp_ep_t ep;
+        sock_dtls_session_get_udp_ep(&socket.ctx_dtls_session, &ep);
+
+        /* Remove all memos of the concerned session. TODO: oberservable memos! */
+        for (int i = 0; i < CONFIG_GCOAP_REQ_WAITING_MAX; i++) {
+            if (_coap_state.open_reqs[i].state == GCOAP_MEMO_UNUSED) {
+                continue;
+            }
+            gcoap_request_memo_t *memo = &_coap_state.open_reqs[i];
+            if (sock_udp_ep_equal(&memo->remote_ep, &ep)) {
+                _expire_request(memo);
+                event_timeout_clear(&memo->resp_evt_tmout);
+            }
+        }
+    }
+
+    if (type & SOCK_ASYNC_MSG_RECV) {
+        ssize_t res = sock_dtls_recv(sock, &socket.ctx_dtls_session, _listen_buf,
+                                    sizeof(_listen_buf), 0);
+        if (res <= 0) {
+            DEBUG("gcoap: DTLS receive failure: %d\n", (int)res);
+            return;
+        }
+        sock_udp_ep_t ep;
+        sock_dtls_session_get_udp_ep(&socket.ctx_dtls_session, &ep);
+        _process_coap_pdu(&socket, &ep,  _listen_buf, res);
+    }
+}
+#endif
+
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+/* Timeout function to free up a session when too many session slots are occupied */
+static void _dtls_free_up_session(void *arg) {
+    (void)arg;
+    sock_dtls_session_t session;
+
+    uint8_t minimum_free = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS;
+    if (dsm_get_num_available_slots() < minimum_free) {
+        if(dsm_get_oldest_used_session(&_sock_dtls, &session) != -1) {
```

Comment:
since this is the only use of this function, why not make it `dsm_drop_oldest_used_session()` and save yourself the need for iterating the sessions array twice?

Reply:
You mean: Let `dsm` destroy the session instead of getting it and then destroy?

Reply:
Yes, if the only reason to get the oldest session is to destroy it, it might as well be done internally.

Reply:
My concern is that it moves a decision from the user to dsm. `sock_dtls_session_destroy()` is one part of the DTLS API which needs a refactor. When ignoring my currently applied tinyDTLS hotfix: `sock_dtls_session_destroy()` can fail. It merely initiates session destroy. But how do handle closing failure? Re-try? Force destroy? I'd argue that should decide the user per application. More details on this topic is in my PR for this: https://github.com/RIOT-OS/RIOT/pull/16422

That's why I avoided implementing any real session handling that concerns the DTLS stack in dsm.